### ISSUE TITLE
SPIN Construct Rule Fails To Run Upon Instantiation when Same SPARQL Query Runs in Workbench

### ISSUE THREAD

**Issue by gwcox66** [eclipse/rdf4j#213]
I apologize for the complexity of this post.  I'm trying to make the problem reproducible.  I'm using Sesame Open RDF Workbench: 

```
System Information
Application Information
Application Name	OpenRDF Workbench
Version	4.1.2
Runtime Information
Operating System	Windows 8.1 6.3 (amd64)
Java Runtime	Oracle Corporation Java HotSpot(TM) 64-Bit Server VM (1.8.0_91)
Process User	Greg
Memory
Used	155 MB
Maximum	3463 MB

```
I'm having trouble getting a SPIN Rule to execute upon instantiation of a class.  `SXXIComplianceChecy7.rdf` (enclosed via zip) has one SPIN rule (below)
[SXXIComplianceCheck7.zip](https://github.com/eclipse/rdf4j/files/342437/SXXIComplianceCheck7.zip)

```
# WARNING: A SPECIAL HANDLING CODE IS REQUIRED (005 03)
CONSTRUCT {
    ?this sxxicc:pub7ProposalHasComplianceMessage "# WARNING: A SPECIAL HANDLING CODE IS REQUIRED (005 03)"^^xsd:string .
}
WHERE {
    ?this sxxicc:pub7ProposalHasDataItem ?dataItem005 .
    ?dataItem005 a sxxicc:Pub7DataItem005 .
    ?dataItem005 sxxicc:pub7DataItemHasStringValue ?securityString .
    ?this sxxicc:pub7ProposalHasDataItem ?dataItem102 .
    ?dataItem102 a sxxicc:Pub7DataItem102 .
    ?dataItem102 sxxicc:pub7DataItemHasStringValue ?serialNumString .
    BIND (fn:string-length(?securityString) AS ?securityStrlen) .
    BIND (SUBSTR(?securityString, 2, 1) AS ?specialHandlingCode) .
    BIND (SUBSTR(?serialNumString, 1, 4) AS ?orgString) .
    FILTER ((?securityStrlen = 1) || (?specialHandlingCode = " "^^xsd:string)) .
    FILTER (?orgString = "AF  "^^xsd:string) .
}

```
I attempt to test this by loading `SXXIComplianceCheck7.rdf` into the default context/graph using the workbench Modify/Add function.  I then execute the following SPARQL Update query to instantiate an instance of `sxxicc:Pub7Proposal` that should cause this SPIN rule to assert the triple in the `CONSTRUCT `clause above (a warning message).

```
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX sxxicc: <http://www.disa.mil/dso/a2i/ontologies/PBSM/Interface/SXXIComplianceCheck#>
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX sp: <http://spinrdf.org/sp#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX smf: <http://topbraid.org/sparqlmotionfunctions#>
PREFIX fn: <http://www.w3.org/2005/xpath-functions#>
PREFIX spl: <http://spinrdf.org/spl#>
PREFIX spin: <http://spinrdf.org/spin#>
PREFIX arg: <http://spinrdf.org/arg#>
PREFIX SXXIComplianceCheckIndividuals: <http://www.disa.mil/dso/a2i/ontologies/PBSM/Interface/SXXIComplianceCheckIndividuals#>
PREFIX sxxicci: <http://www.disa.mil/dso/a2i/ontologies/PBSM/Interface/SXXIComplianceCheckIndividuals#>
PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>

INSERT DATA
{
   sxxicci:TestPub7Proposal a sxxicc:Pub7Proposal ;
           sxxicc:pub7ProposalHasDataItem sxxicci:testPub7Proposal_DataItem005 ;
           sxxicc:pub7ProposalHasDataItem sxxicci:testPub7Proposal_DataItem017 ;
           sxxicc:pub7ProposalHasDataItem sxxicci:testPub7Proposal_DataItem102 .
 
   sxxicci:testPub7Proposal_DataItem005 a sxxicc:Pub7DataItem005 ;
           sxxicc:pub7DataItemHasStringValue "S"^^xsd:string .
 
   sxxicci:testPub7Proposal_DataItem102 a sxxicc:Pub7DataItem102 ;
           sxxicc:pub7DataItemHasStringValue "AF  881234"^^xsd:string .
 
   sxxicci:testPub7Proposal_DataItem017 a sxxicc:Pub7DataItem017 ;
           sxxicc:pub7DataItemHasStringValue "U"^^xsd:string .
}
```

However, after executing this and exploring the `sxxicci:TestPub7Proposal`, I see that it has no warning message.  Specifically, I explore by using the **same** construct query (copy/paste) from the workbench Explore/Query (`PREFIX`'s as shown for the `UPDATE `above).  The query produces the expected triple:

`
sxxicci:TestPub7Proposal     sxxicc:pub7ProposalHasComplianceMessage   "# WARNING: A SPECIAL HANDLING CODE IS REQUIRED (005 03)"
`

So, this triple **should** have been asserted by SPIN upon instantiation of `sxxicci:TestPub7Proposal`, However, when I click into/explore `sxxicci:TestPub7Proposal`, I don't see the triple above, but all the remaining information for the instance is as it should be.

I started to pare down my SPIN rule query.  In version 8 of my RDF file (enclosed below via ZIP), I only removed the last `FILTER `statement from my SPIN rule query.  I repeated the methodology above, and the desired warning message triple WAS automatically asserted by the SPIN rule upon instantation of my` sxxicci:TestPub7Proposal`.  I'm confused.  I do need that last filter statement.  Is this a bug or am I doing something wrong?  Help?  Thanks!

[SXXIComplianceCheck8.zip](https://github.com/eclipse/rdf4j/files/342476/SXXIComplianceCheck8.zip)

**Comment by gwcox66** [comment]
One possibly-troublesome feature of the SPIN rule in this test case is that the second `BIND `will have trouble since the test case has a `?securityString` of length 1.  The second `BIND `in the SPIN rule tries to get the second character of the `?securityString`, which doesn't exist.  However, this issue is present in both version 7 and version 8 of my RDF file and in the manually-run queries. I would think that the second bind should assign an empty string to `?specialHandlingCode`.

**Comment by pulquero** [comment]
I have confirmed in the case of the first filter that substr experiences an indexoutofboundsexception as you suspected, which causes it not to be bound.
The second filter fails because in spin the RHS is the literal "AF"^^xsd:string, the trimmed value missing the trailing double space. I'm gonna blame the rdf xml loader because the triple in the repo already has the trimmed value. Try using another format like turtle.

**Comment by pulquero** [comment]
Yep, the error is on line 398, SAXFilter. It trims "AF  " to "AF".

**Comment by gwcox66** [comment]
Thanks @pulquero! This information will help me work around this.  I'll see if I can use another format OR see if I can avoid literal strings in my rules that would trim to something I don't intend.

**Comment by gwcox66** [comment]
@pulquero has addressed this in Issue #219.  I also have an (ugly) work-around that avoids the trim issue, allowing my SPIN rule query to run by extending my literal string to avoid trailing spaces that would be trimmed, then manually removing the eXtra character using a `SUBSTR` operation (second half of second `FILTER `statment):

```
# WARNING: A SPECIAL HANDLING CODE IS REQUIRED (005 03)
CONSTRUCT {
    ?this sxxicc:pub7ProposalHasComplianceMessage "# WARNING: A SPECIAL HANDLING CODE IS REQUIRED (005 03)"^^xsd:string .
}
WHERE {
    ?this sxxicc:pub7ProposalHasDataItem ?dataItem005 .
    ?this sxxicc:pub7ProposalHasDataItem ?dataItem102 .
    ?dataItem005 a sxxicc:Pub7DataItem005 .
    ?dataItem005 sxxicc:pub7DataItemHasStringValue ?securityString .
    ?dataItem102 a sxxicc:Pub7DataItem102 .
    ?dataItem102 sxxicc:pub7DataItemHasStringValue ?serialNumString .
    FILTER ((fn:string-length(?securityString) = 1) || (SUBSTR(?securityString, 2, 1) = " "^^xsd:string)) .
    FILTER (SUBSTR(?serialNumString, 1, 4) = SUBSTR("AF  X"^^xsd:string, 1, 4)) .
}
```
